airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/metadata_assets.py

```diff
@@ -0,0 +1,182 @@
+import pandas as pd
+from dagster import Output, asset
+
+from metadata_service.models.generated.ConnectorMetadataDefinitionV1 import ConnectorMetadataDefinitionV1
+from ..utils.object_helpers import are_values_equal, merge_values
+from ..utils.dagster_helpers import OutputDataFrame
+
+GROUP_NAME = "metadata"
+
+
+# HELPERS
+
+OSS_SUFFIX = "_oss"
+CLOUD_SUFFIX = "_cloud"
+
+
+def get_primary_catalog_suffix(merged_df):
+    """
+    Returns the suffix for the primary catalog and the secondary catalog.
+    The primary catalog is the one that is used for the final metadata.
+    The secondary catalog is the one that is used for overrides.
+
+    This is nessesary because we have connectors that are only in the cloud catalog and vice versa.
+    """
+    cloud_only = merged_df["_merge"] == "right_only"
+    primary_suffix = CLOUD_SUFFIX if cloud_only else OSS_SUFFIX
+    secondary_suffix = OSS_SUFFIX if cloud_only else CLOUD_SUFFIX
+    return primary_suffix, secondary_suffix
+
+
+def get_field_with_fallback(merged_df, field):
+    """
+    Returns the value of the field from the primary catalog.
+    If the field is not present in the primary catalog, the value from the secondary catalog is returned.
+    """
+
+    primary_suffix, secondary_suffix = get_primary_catalog_suffix(merged_df)
+
+    primary_field = field + primary_suffix
+    secondary_field = field + secondary_suffix
+
+    secondary_value = merged_df.get(secondary_field)
+    return merged_df.get(primary_field, default=secondary_value)
+
+
+def compute_catalog_overrides(merged_df):
+    """
+    Returns the catalog overrides section for the metadata file.
+    """
+    cloud_only = merged_df["_merge"] == "right_only"
+    oss_only = merged_df["_merge"] == "left_only"
+
+    catalogs = {
+        "oss": {
+            "enabled": not cloud_only,
+        },
+        "cloud": {
+            "enabled": not oss_only,
+        },
+    }
+
+    if cloud_only or oss_only:
+        return catalogs
+
+    allowed_overrides = [
+        "name",
+        "dockerRepository",
+        "dockerImageTag",
+        "supportsDbt",
+        "supportsNormalization",
+        "license",
+        "supportUrl",
+        "connectionType",
+        "allowedHosts",
+        "normalizationConfig",
+        "suggestedStreams",
+        "resourceRequirements",
+    ]
+
+    # find the difference between the two catalogs
+    for override_col in allowed_overrides:
+        oss_col = override_col + OSS_SUFFIX
+        cloud_col = override_col + CLOUD_SUFFIX
+
+        cloud_value = merged_df.get(cloud_col)
+        oss_value = merged_df.get(oss_col)
+
+        # if the columns are different, add the cloud value to the overrides
+        if cloud_value and not are_values_equal(oss_value, cloud_value):
+            catalogs["cloud"][override_col] = merge_values(oss_value, cloud_value)
```

Comment:
If I understand it correcty, this part checks whether, for an identical list of connectors on cloud and connectors on OSS, if the columns are different, and if so can perform overrides based on column differences (e.g. pinning a version). 

If we have connectors that are only on oss or cloud, it looks like we will exit early on L63 - do we then miss out on this column-based override?

Reply:
Good eye Ella!

I think this logic is still correct. But it always felt like a scratchy sweater. So let me lay out some background context, then how this is meant to work.

**Types of catalog entries**
There are four scenarios to know about for a connector as it relates to a catalog
1. available in oss but not cloud
2. available in oss and cloud 
3. not available in oss but available in cloud
4. not available in either (we can ignore this for now)

**How they relate to overrides**
Scenario 1 its straight forward. All oss values are declared under `data` and `data.catalogs.oss.enabled` is `true`

Scenario 2 is also straight forward. All oss values are declared under `data`, `data.catalogs.oss.enabled` is `true`, `data.catalogs.cloud.enabled` is `true`, and any cloud specific values is under `data.catalogs.cloud`

Scenario 3 is straight forward to say, but the logic in the code feels funny. All cloud values are declared under `data` and `data.catalogs.cloud.enabled` is `true`. So no overrides occur, because its only in one catalog

**Onto the code**
Ok so looking at the code were commenting on.
1. If its a cloud or oss only connector, we can skip overrides. 
2. For differences in field values between the connector definition in cloud vs oss, the cloud value should be placed under `data.catalogs.cloud`

Does that make sense? Is there an edge case im missing?

oh and also for reference there is a [child PR](https://github.com/airbytehq/airbyte/pull/24545) to this one that diffs the new catalogs with the old one